receiver/splunkhecrawreceiver/receiver.go

```diff
@@ -0,0 +1,263 @@
+		return nil, errNilNextLogsConsumer
+	}
+
+	if config.Endpoint == "" {
+		return nil, errEmptyEndpoint
+	}
+
+	r := &splunkReceiver{
+		logger:       logger,
+		config:       &config,
+		logsConsumer: nextConsumer,
+		server: &http.Server{
+			Addr: config.Endpoint,
+			// TODO: Evaluate what properties should be configurable, for now
+			//		set some hard-coded values.
+			ReadHeaderTimeout: defaultServerTimeout,
+			WriteTimeout:      defaultServerTimeout,
+		},
+	}
+
+	return r, nil
+}
+
+// splunkReceiver implements the component.LogsReceiver for Splunk HEC raw protocol.
+type splunkReceiver struct {
+	logger             *zap.Logger
+	config             *Config
+	logsConsumer       consumer.Logs
+	server             *http.Server
+	obsrecv            *obsreport.Receiver
+	resourceCustomizer func(*http.Request, pdata.Resource)
+}
+
+// Start tells the receiver to start its processing.
+// By convention the consumer of the received data is set when the receiver
+// instance is created.
+func (r *splunkReceiver) Start(_ context.Context, host component.Host) error {
+	var ln net.Listener
+	// set up the listener
+	ln, err := r.config.HTTPServerSettings.ToListener()
+	if err != nil {
+		return fmt.Errorf("failed to bind to address %s: %w", r.config.Endpoint, err)
+	}
+
+	mx := mux.NewRouter()
+	mx.NewRoute().HandlerFunc(r.handleReq)
+
+	r.server = r.config.HTTPServerSettings.ToServer(mx)
+
+	// TODO: Evaluate what properties should be configurable, for now
+	//		set some hard-coded values.
+	r.server.ReadHeaderTimeout = defaultServerTimeout
+	r.server.WriteTimeout = defaultServerTimeout
+	r.resourceCustomizer = r.createResourceCustomizer()
+
+	go func() {
+		if errHTTP := r.server.Serve(ln); errHTTP != http.ErrServerClosed {
+			host.ReportFatalError(errHTTP)
+		}
+	}()
+
+	return err
+}
+
+// Shutdown tells the receiver that should stop reception,
+// giving it a chance to perform any necessary clean-up.
+func (r *splunkReceiver) Shutdown(context.Context) error {
+	return r.server.Close()
+}
+
+func (r *splunkReceiver) handleReq(resp http.ResponseWriter, req *http.Request) {
+	ctx := req.Context()
+	if r.logsConsumer == nil {
+		ctx = r.obsrecv.StartMetricsOp(ctx)
+	}
+	reqPath := req.URL.Path
+	if !r.config.pathGlob.Match(reqPath) {
+		r.failRequest(ctx, resp, http.StatusNotFound, notFoundRespBody, nil)
+		return
+	}
+
+	if req.Method != http.MethodPost {
+		r.failRequest(ctx, resp, http.StatusBadRequest, invalidMethodRespBody, nil)
+		return
+	}
+
+	encoding := req.Header.Get(httpContentEncodingHeader)
+	if encoding != "" && encoding != gzipEncoding {
+		r.failRequest(ctx, resp, http.StatusUnsupportedMediaType, invalidEncodingRespBody, nil)
+		return
+	}
+
+	bodyReader := req.Body
+	if encoding == gzipEncoding {
+		var err error
+		bodyReader, err = gzip.NewReader(bodyReader)
+		if err != nil {
+			r.failRequest(ctx, resp, http.StatusBadRequest, errGzipReaderRespBody, err)
+			return
+		}
+	}
+
+	if req.ContentLength == 0 {
```

Comment:
It seems more obvious to put this above around line 166 (we don't need to ungzip if content length is 0 right?)

Reply:
good one.